### ISSUE TITLE
Failed creating .X11-unix and .ICE-unix with proper permissions

### ISSUE THREAD

**Issue by troglobit** [troglobit/finit#46]
From issue #42 @xhebox reports:

 I've noticed that `.X11-unix` is not `1777`, the same problems about `.ICE-unix` and `makedir(xxx)`?Also, there's a `makedir("/var/lock/subsys"0755);` in `bootmisc-plugin`, then why we get error before patched  `dbus-plugin`?

UPDATE2: @troglobit http://stackoverflow.com/questions/8719867/linux-mkdir-function-cant-authorize-full-permission. It looks like `umask`(affect `mkdir` but `chmod`)...
```
#include <sys/stat.h>

int main() {
        printf("%d\n", mkdir("/tmp/s", 777));
        return 0;
}

dr----x--t  2 xhe  users 4096 Oct  2 13:10 s
```
```
#include <sys/stat.h>

int main() {
        printf("%d\n", mkdir("/tmp/s", 777));
        printf("%d\n", chmod("/tmp/s", 01777));
        return 0;
}
drwxrwxrwt  2 xhe  users 4096 Oct  2 13:10 s
```
So your `lite` lib needs to be fixed like(also other functions about `perm`): 
```
static inline int makedir(const char *path, mode_t mode)
{
        if (mkdir(path, mode) && errno != EEXIST)
                return -1;
        if (chmod(path, mode))
                return -1;
        return 0;
}
```

**Comment by xhebox** [comment]
Wait...Should not this be in libite?

**Comment by troglobit** [comment]
Maybe, either way it needs to be followed up here as well. Would need to
reference a new libite version at least.

I'm just about to look into it :)

**Comment by troglobit** [comment]
@xhebox The correct fix, as mentioned in the StackOverflow question you linked to, is to set umask() properly before calling `mkdir()` et al.  

I've tested the following fix in [TroglOS](https://github.com/troglobit/troglos), and it works. Don't understand what "op" did wrong in the StackOverflow question ...

```diff
diff --git a/plugins/bootmisc.c b/plugins/bootmisc.c
index 30d7a7f..225c6d1 100644
--- a/plugins/bootmisc.c
+++ b/plugins/bootmisc.c
@@ -43,6 +43,8 @@ static void setup(void *UNUSED(arg))
 {
        int gid;
 
+       umask(0);
+
        _d("Setting up FHS structure in /var ...");
        makedir("/var/cache",      0755);
        makedir("/var/games",      0755);
@@ -104,6 +106,8 @@ static void setup(void *UNUSED(arg))
        makedir("/var/run/sshd",  01755); /* OpenSSH  */
        makefifo("/dev/xconsole",  0640); /* sysklogd */
        chown("/dev/xconsole", 0, getgroup("tty"));
+
+       umask(022);
 }
 
 static plugin_t plugin = {
diff --git a/plugins/x11-common.c b/plugins/x11-common.c
index 9b9cbe9..1463533 100644
--- a/plugins/x11-common.c
+++ b/plugins/x11-common.c
@@ -37,6 +37,7 @@ static void setup(void *UNUSED(arg))
 #endif
        char line[LINE_SIZE];
 
+       umask(0);
        makedir("/var/run/console", 01777);
        snprintf(line, sizeof(line), "/var/run/console/%s", username);
        touch(line);
@@ -54,6 +55,8 @@ static void setup(void *UNUSED(arg))
 
        if (fexist("/sbin/restorecon"))
                run("/sbin/restorecon /tmp/.ICE-unix /tmp/.X11-unix");
+
+       umask(022);
 }
 
 static plugin_t plugin = {
```

**Comment by xhebox** [comment]
@troglobit hmm..Set umask is indeed right, but you'll need to set correct umask every time use the function...Also `mkfifo` is affected.
I've searched your lib, this is the list of affected functions:
```
makepath
makedir
makefifo
mdir
```
Also, `tests/rsync.c` calls mkdir, too.

**Comment by troglobit** [comment]
Yup, every time.  Which I think is unavoidable since umask is a security measure.

We don't want to leave umask as 0 and we don't want to wrap the umask in the library, because use-cases differ ... 

Good catch, I'll add an issue to libite to update the tests and the README with a notice on the umask.

**Comment by xhebox** [comment]
What about `makedir("/var/lock/subsys",0755);`?Both dbus and bootmisc call it(you commited with my first patch).Is there a need of removing?

**Comment by troglobit** [comment]
Didn't understand first what you meant, then I too noticed the duplicate `makedir()` of `/var/lock/susbys`, thanks! :+1: 

Fixed in 4eb5f8a :smiley:
